api/services/BuildResolver.js

```diff
@@ -0,0 +1,73 @@
+const { Build, Site } = require('../models');
+const buildErrors = require('../responses/buildErrors');
+const GitHub = require('../services/GitHub');
+
+const rejectBuild = msg =>
+  Promise.reject({
+    status: 404,
+    message: msg,
+  });
+
+const getBranch = ({ user, site, branch }) =>
+  GitHub.getBranch(user, site.owner, site.repository, branch)
+  .then(branchInfo => ({
+    branch: branchInfo.name,
+    site: site.id,
+    commitSha: branchInfo.commit.sha,
+  }))
+  .catch(() => rejectBuild(buildErrors.BRANCH_NOT_FOUND));
+
+const getBuildById = (user, params) => {
+  const { buildId, siteId } = params;
+
+  return Build.findOne({
+    where: {
+      id: buildId,
+      site: siteId,
+    },
+  })
+  .then((model) => {
+    if (model) {
+      return model;
+    }
+
+    return rejectBuild(buildErrors.BUILD_NOT_FOUND);
+  });
+};
+
+const getBuildByBranch = (user, params) => {
+  const { siteId, sha, branch } = params;
+
+  return Site.findOne({
+    where: { id: siteId },
+    attributes: ['id', 'owner', 'repository'],
+    include: {
+      model: Build,
+      where: { branch },
+      required: false,
+    },
+  })
+  .then((model) => {
+    const site = model;
+    const build = site.Builds && site.Builds[0];
+
+    // The branch we want to create a new build from has been built via federalist before
+    if (build) {
+      return Object.assign({}, build.toJSON(), { commitSha: sha });
```

Comment:
the `build` should/can already have a `commitSha`, right? Do we really want to overwrite it here if it does?

Reply:
It isn't guaranteed to have one, no. I have a few builds on staging at least that were completed successfully but don't have SHAs for whatever reason. We can not overwrite it if present but we do need to account for it.

If this isn't an issue in the prod database I'm happy to remove it.

Reply:
It is probably the case that some builds in prod are missing it as well since that field was added in a migration in 2017.

Maybe just add it if it isn't there already?

```js
const serializedBuild = build.toJSON();
serializedBuild.commitSha = serializedBuild.commitSha || sha;
return serializedBuild;
```